README.md
# Express Mustache Overlays

Serves mustache templates and partials, checking each directory in turn for matches, and reloading on file changes.


## Configuration

The components in this package make use of the `app.locals.mustache` namespace. The `prepareMustache()` function helps set up the data structure correctly.

Configuration environment variables for the example.

* `MUSTACHE_DIRS` - A `:`-separated list of directories to check for templates e.g. `mustache-overlay:mustache`.

**Any configuration from `MUSTACHE_DIRS` gets merged into existing configuration such that it is used in preference to it. Effectively, the `MUSTACHE_DIRS` settings override settings defined in code.**

Additionally:

* `DEBUG` - Include `express-mustache-overlays` to get debug output from the `express-mustache-overlays` library itself and `express-mustache-overlays:server` for messages from the example server. Also include `express-public-files-overlays` to get debug from the public files server in the example.


## Internal Workings

Internally, the code is designed to work in these stages:

* `mustacheFromEnv(app)` - Parses and returns the config from the `MUSTACHE_DIRS` environment variable
* `prepareMustache(app, userDirs)` - Sets up the `app.locals.mustache` data structure with a `userDirs` and a `libDirs` key and makes `app.locals.mustache.overlay()` available (see next). The `userDirs` argument is optional. You usually pass the output of `mustacheFromEnv(app)` as the `userDirs` variable. Any library directories (that should be used if a match can't be found in the `userDirs`) can be set up using `overlay()` described next to add them into `libDirs`.
* `app.locals.mustache.overlay(dirs)` - A function other libraries can use to merge any overlays they need into the `libDirs` configuration. The `userDirs` configuration will always overlay over the `libDirs` configuration, even if it is set up earlier.
* `setupMustache(app)` - Installs the middleware based on the settings in `app.locals.mustache`. This should always come last.

Internally a watch is set up using `chokidar` on any partials that are present
so that the updated contents can be used by the templates. Since templates (but
not partials) are read each time they are rendered, the watches are only added
to the partials since the latest template content will be rendered anyway.

### Accessing the Overlays Object

You can access the overlays object like this once `prepareMustache()` is called:

```
app.locals.mustache.overlaysPromise.then((overlays) => {
  // Use overlays here
})
```

Ordinarily you wouldn't need this, but the object can be useful if you want to use the template system outside of Express.

The `overlays` object from the promise has these methods:

* `findView(template)` - async function (requires `await` when called) which resolves to the path on the filesystem of the view
* `renderView(template, options)` - async function (requires `await` when called) which resolves to the template named `template`, rendered with `options`. E.g. `const html = await renderView('content', {content: 'hello'})`
* `renderFile(path, options)` - async function (requires `await` when called) which resolves takes the `path` as the full path to the mustache template, and the same `options` as `renderView()`.


## Example

```
const express = require('express')
const path = require('path')
const { prepareMustache, setupMustache, mustacheFromEnv } = require('../index.js')

const app = express()
prepareMustache(app, mustacheFromEnv(app))

// Any other express setup can change app.locals.mustache.libDirs here to add
// additional library-defined public files directories to be served.  Any user
// defined directories will be prepended before any corresponding URL path in
// the library directories list. The safest way to add overlays is with the
// overlay() function demonstrated here.
app.locals.mustache.overlay([path.join(__dirname, 'mustache')])

// Add any routes here:
app.get('', (req, res) => {
  res.render('hello', {})
})

// Set up the engine
const mustacheEngine = setupMustache(app)
app.engine('mustache', mustacheEngine)
app.set('views', app.locals.mustache.dirs)
app.set('view engine', 'mustache')

app.listen(8000, () => console.log(`Example app listening on port 8000`))
```

The `mustache` directory contains a `hello.mustache` template.

See the `./example` directory for an example.

```
cd example
npm install
```

Then follow the instructions in the `README.md` in the `example` directory.


## Dev

```
npm run fix
```


## Changelog

### 0.5.3 2019-02-15

* Doc fixes
* Changed `prepareMustache = (app, userDirs, libDirs)` -> `prepareMustache = (app, userDirs)`
* Added a warning if directories don't exist

### 0.5.2 2019-02-07

* Don't throw an error in the renderer, it creates an `UnhandledPromiseRejectionWarning`. Calling the callback with an error is enough?
* Improved the Docker example
* Improved logging

### 0.5.1 2019-02-07

* Changed debug behaviour to use own `debug()`, not `app.locals.debug()`.
* Moved the example to `./example`.

### 0.5.0 2019-02-06

* Complete refactor. See old `CHANGELOG.md` for older changes. Removed all functionality apart from the overlays behaviour. See `express-public-files-overlays` for static file serving.


CHANGELOG.md
# Changelog

**Note: More recent changes can be found in `README.md`.**

## 0.4.3 2019-01-19

* Added a `pjaxHandlers.mustache` partial

## 0.4.2 2019-01-19

* Added `DEFAULT_TITLE` environment variable which sets the `title` option
* Set a debug error message if the template rendering fails

## 0.4.1 2019-01-18

* Added `SIGTERM` and `SIGNINT` handlers

## 0.4.0 2019-01-18

* Add Dockerfile
* Refactored to make the `overlays` object more useful, have more re-usable code in `lib/index.js` rather than in `bin/server.js`
* Renamed `publicURLPath` to `publicUrlPath`
* Added `sharedPublicURLPath` and used it to replace `publicUrlPath` in default templates
* Renamed `offlineUrl` to `networkErrorUrl`, `OFFLINE_URL` to `NETWORK_ERROR_URL` and `/offline` to `/network-error`
* Return of `renderFile`, `renderView` and `findView` from `overlays.setup()` is deprecated. Just use them as methods on `overlays` object.
* Make `networkError.mustache` and `start.mustache`
* Make `/`, `/ok` and `/throw`  optional, defaulting to not present, but enabled by setting the environment variable `DEMO_ROUTES=true`

## 0.3.9 2019-01-12

* Added `THEME_COLOR` environment variable
* Added an `Install App` link which appears instead of Chrome Android automatically asking the user

## 0.3.8 2019-01-12

* Added `overlaysOptionsFromEnv()` function to parse all the overlays options that can be specified as environment variables.
* Added `overlaysDirsFromEnv()` function to parse the `MUSTACHE_OVERLAYS_DIRS` and `PUBLIC_FILES_DIRS` environment variables.

## 0.3.7 2019-01-12

* Some more partials JS fixes **Caution: Note that `serviceWorkerUrl` and `offlineUrl` are not escaped in the template, so make sure `SERVICE_WORKER_URL` and `OFFLINE_URL` are trusted.**

## 0.3.6 2019-01-12

* Ensured `renderView()` gets all the variables from `app.locals`
* Added `OFFLINE_URL`, `MANIFEST_URL`, `SERVICE_WORKER_URL` and `ICON_192_URL` config options
* Refactored `top.mustache` as well as the scripts partials to use the new variables

## 0.3.5 2019-01-12

* Made the default template behave as a single page app using PJAX and a service worker, and added URLs that are needed to serve this as a progressive web app via `gateway-lite`
  * Added a wrapper `pjax-container` to the entire body of the template
  * Changed from jQuery slim to normal jQuery
  * Introduced jQuery PJAX so that full page reloads aren't required as you navigate, instead the `<div id="pjax-container">` content is replaced into the existing div. **Caution: This means extra infomtation in the `<head>` or reloaded outside `<div id="pjax-container">` will not be loaded
  * Added PJAX initialisation that installs a `pjax:error` handler which inserts a simple you are offline message
  * Added a `/offline` route to `bin/server.js` to generate a simple offline message page
  * Changed `<meta name="apple-mobile-web-app-capable" content="yes">` to `no` since apple still doesn't handle PWAs very well (cookies are lost as users switch to a different app and back).
  * Removed `manifest.json` and the icons as they can now be served by gateway-lite instead
  * WONT Support PJAX container-only HTML responses - makes the cacheing of the responses by service worker more complex.
  * Changed partial structure so that there are more overrides possible
  * Added gzip compresssion to the example
  * Setting the `metaDescription` local to a string results in a meta description tag being added for SEO
  * Make the PJAX offline message the same as the one in the service worker cache by fetching that page with an AJAX call.
  * Add online/offline handlers as a partial
  * Create a `/start` URL that is used when the app starts.
  * By default the service worker install, uninstall, pjax and install prompt are all disabled in `views/partials/bodyEnd.mustache` with the use of the `!` (a mustache comment) in the partial includes.

## 0.3.4 2019-01-04

* Added a `options` to `setupErrorHandlers()` to allow `debug` to optionally be passed to the error handlers. This means errors from an app will be logged with the apps they come from, rather than from `express-mustache-overlays`.

## 0.3.3 2018-12-29

* Return `{renderFile, findView, renderView}` from `setup()`, not just the `renderFile` function. The new `renderView` function will resolve the correct template file based on its name (just like calls to `res.render()` so that you don't need to do the looking up yourself.
* Added missing `debug` import in `bin/server.js`

## 0.3.2 2018-12-29

* Wrapped the content in the `views/content.mustache` template in an `<article>` tag.

## 0.3.1 2018-12-29

* `overlays.setup()` now returns a `render(path, options)` method that can also be used outside the Express context

## 0.3.0 2018-12-21

* Split `setupMustacheOverlays(...)` into `const overlays = prepareMustacheOverlays(...)` and `overlays.setup()`
* Refactored to use `app.locals` for `title`, `publicURLPath` and `scriptName`
* Shortened the watch throttle timeout on changed partials from 500ms to 200ms
* Prevented partials reload events happening on initial load
* Added error checking from shelljs command

## 0.2.2 2018-12-20

* Changed the way environment variables are handled internally. They are used in `bin` but not in `lib` now.

## 0.2.1 2018-12-20

* Removed 403 template. Makes more sense in `express-mustache-jwt-signin`
* Refactored `bottom.mustache` to use `footer.mustache`

## 0.2.0 2018-12-20

* Refactored to support public files too as well as to host 400, 403 and 500 pages

## 0.1.4 2018-12-15

* Tweaked example to set from `MUSTACHE_DIRS` correctly

## 0.1.3 2018-12-15

* Support the `MUSTACHE_DIRS` environment variable as a `:` separated list of paths to look at first before looking in the default `views` directory
* Listen to `all` events in partials such as add and delete as well as change
* Improved logging of which partials directory is being reloaded
* Throttle reloading with `lodash._throttle` and support sub directories


lib/index.js
const mustache = require('mustache')
const path = require('path')
const debug = require('debug')('express-mustache-overlays')
const fs = require('fs')
const shelljs = require('shelljs')
const chokidar = require('chokidar')
const _ = require('lodash')
const { promisify } = require('util')
const readFileAsync = promisify(fs.readFile)
const accessAsync = promisify(fs.access)

const mustacheFromEnv = (app) => {
  const dirs = process.env.MUSTACHE_DIRS ? process.env.MUSTACHE_DIRS.split(':') : []
  debug('MUSTACHE_DIRS environment variable directories:', dirs)
  return dirs
}

const loadPartials = async (mustacheDirs) => {
  const partials = {}
  if (!Array.isArray(mustacheDirs)) {
    throw Error('Expected an array of paths for mustachDirs')
  }
  mustacheDirs.reverse()
  for (let mustacheDir of mustacheDirs) {
    // Make it the same format that shelljs returns
    mustacheDir = path.normalize(mustacheDir)
    const partialsPath = path.join(mustacheDir, 'partials', '*.mustache')
    const files = shelljs.ls(partialsPath)
    if (shelljs.error()) {
      debug(`Error listing mustache partial files matching '${partialsPath}', perhaps none were found`)
    }
    if (!files.length) {
      debug('No partials found. Continuing anyway ...')
    }
    for (const filename of files) {
      const name = filename.slice((mustacheDir + '/partials/').length, filename.length - 9)
      partials[name] = await readFileAsync(filename, { encoding: 'utf8' })
    }
  }
  mustacheDirs.reverse()
  // Should match the file extension of the view.
  return partials
}

const prepareMustache = (app, userDirs) => {
  app.locals.mustache = {
    userDirs: userDirs || [],
    libDirs: [],
    overlay: (dirs) => {
      // Safer to use the whole path in case someone has replaced
      // the objects
      for (let i = 0; i < dirs.length; i++) {
        const dir = path.normalize(dirs[i])
        accessAsync(dir, fs.constants.R_OK).catch((e) => {
          const msg = `WARNING: Cannot access directory ${dir}: ${e}`
          console.error(msg)
          debug(msg)
          return true
        })
        debug(`Overlaying the directory ${dir} into app.locals.mustache.libDirs`)
        app.locals.mustache.libDirs.unshift(dir)
      }
    }
  }
}

const setupMustache = (app) => {
  let mustacheDirs = []
  mustacheDirs = mustacheDirs.concat(app.locals.mustache.userDirs)
  mustacheDirs = mustacheDirs.concat(app.locals.mustache.libDirs)
  app.locals.mustache.dirs = mustacheDirs
  const overlaysPromise = new Promise(async (resolve, reject) => {
    debug('Setting up templates...')
    debug('Using these mustache views directories in order of preference:', mustacheDirs)
    let partials = await loadPartials(mustacheDirs)
    const partialsDirs = []
    const partialsDirsMustacheGlobs = []
    for (let mustacheDir of mustacheDirs) {
      partialsDirs.push(path.join(mustacheDir, 'partials'))
      partialsDirsMustacheGlobs.push(path.join(mustacheDir, 'partials', '**', '*.mustache'))
    }
    // Look for changes to partials directories, and reload if needed
    debug('Watching these globs:', partialsDirsMustacheGlobs.join(','))
    const onEvent = async (event, path) => {
      debug(`Reloading partials due to ${event} on ${path} ...`)
      partials = await loadPartials(mustacheDirs)
    }
    const throttledEvent = _.throttle(onEvent, 200, { 'trailing': true, 'leading': false })
    chokidar.watch(partialsDirsMustacheGlobs, { ignoreInitial: true, ignored: /(^|[/\\])\../ }).on('all', throttledEvent)
    const renderFile = async (filePath, options) => {
      const content = await readFileAsync(filePath, { encoding: 'utf8' })
      const rendered = mustache.render(content, Object.assign({}, options), partials)
      return rendered
    }
    const findView = async (template) => {
      let templatePath
      for (let i = 0; i < mustacheDirs.length; i++) {
        const filePath = path.join(mustacheDirs[i], template + '.mustache')
        try {
          await accessAsync(filePath, fs.constants.R_OK)
          templatePath = filePath
          break
        } catch (e) {
          // Can't access the file
          // debug(e)
        }
      }
      if (!templatePath) {
        throw new Error(`Template '${template}' not found, searched ${mustacheDirs}`)
      }
      return templatePath
    }
    const renderView = async (template, options) => {
      return renderFile((await findView(template)), options)
    }
    const overlays = {
      renderFile,
      renderView,
      findView
    }
    resolve(overlays)
  })
  app.locals.mustache.overlaysPromise = overlaysPromise
  app.locals.mustache.overlaysPromise
    .then((overlays) => {
      debug('Completed setup of mustache overlays.')
    })
  const mustacheEngine = async (filePath, options, callback) => {
    try {
      const overlays = await app.locals.mustache.overlaysPromise
      const rendered = await overlays.renderFile(filePath, options)
      callback(null, rendered)
    } catch (e) {
      debug('Error setting up mustache overlays:', e)
      callback(e, null)
    }
  }
  return mustacheEngine
}

module.exports = { prepareMustache, setupMustache, mustacheFromEnv }


lib/simple.js
const express = require('express')
const path = require('path')
const { optionsFromEnv, mustacheOverlays, mustacheEngine, installSignalHandlers, setupErrorHandlers } = require('./lib.js')


const adjustLocals = (locals, envOptions) => {
  const { mustacheDirs, publicFilesDirs, publicUrlPath = '/public', sharedPublicUrlPath = '/public' } = locals || {}
  if (typeof locals.debug === 'undefined') {
    locals.debug = console.log
  }
  if (typeof publicUrlPath === 'undefined') {
    throw new Error('Please set publicUrlPath')
  }
  if (publicUrlPath.endsWith('/')) {
    throw new Error('The publicUrlPath option must not end with "/"')
  }
  // Example of setting up the main
  const newMustacheDirs = []
  if (mustacheDirs) {
    mustacheDirs.forEach((dir) => {
      newMustacheDirs.push(dir)
    })
  }
  newMustacheDirs.push(path.join(__dirname, '..', 'views'))
  if (envOptions.mustacheDirs) {
    envOptions.mustacheDirs.forEach((dir) => {
      newMustacheDirs.unshift(dir)
    })
  }
  const newPublicFilesDirs = []
  if (publicFilesDirs) {
    publicFilesDirs.forEach((dir) => {
      newPublicFilesDirs.push(dir)
    })
  }
  newPublicFilesDirs.push(path.join(__dirname, '..', 'public'))
  if (envOptions.publicFilesDirs) {
    envOptions.publicFilesDirs.forEach((dir) => {
      newPublicFilesDirs.unshift(dir)
    })
  }
  const newPublicUrlPath = locals.publicUrlPath || envOptions.publicUrlPath || publicUrlPath
  const newSharedPublicUrlPath = locals.sharedPublicUrlPath || envOptions.sharedPublicUrlPath || sharedPublicUrlPath
  const newScriptName = locals.scriptName || envOptions.scriptName
  // Merge new options with the existing ones
  const options = Object.assign({}, locals, envOptions, { publicUrlPath: newPublicUrlPath, mustacheDirs: newMustacheDirs, publicFilesDirs: newPublicFilesDirs, sharedPublicUrlPath: newSharedPublicUrlPath, scriptName: newScriptName })
  if (options.mustacheDirs.length < 1) {
    throw new Error('Expected at least one entry in the mustacheDirs array')
  }
  if (options.scriptName.endsWith('/')) {
    throw new Error('SCRIPT_NAME should not end with /.')
  }
  return options
}

const expressMustacheOverlays = (app, options, callback) => {
  app.locals = adjustLocals(app.locals, optionsFromEnv())
  app.locals.overlaysPromise = mustacheOverlays(app.locals)
  const { debug, startUrl = '/start', networkErrorUrl = '/network-error' } = app.locals

  app.get(startUrl, async (req, res, next) => {
    try {
      res.render('start', { })
    } catch (e) {
      next(e)
    }
  })

  app.get(networkErrorUrl, async (req, res, next) => {
    try {
      res.render('networkError', {})
    } catch (e) {
      next(e)
    }
  })

  callback()

  app.engine('mustache', mustacheEngine(app.locals.overlaysPromise))
  app.set('views', app.locals.mustacheDirs)
  app.set('view engine', 'mustache')
  app.locals.publicFilesDirs.forEach((publicFilesDir) => {
    debug(app.locals.publicUrlPath, '->', publicFilesDir)
    app.use(app.locals.publicUrlPath, express.static(publicFilesDir))
  })

  setupErrorHandlers(app, { debug })
  // Shutdown
  installSignalHandlers()
}

module.exports = { expressMustacheOverlays, adjustLocals }


example/server.js
const express = require('express')
const path = require('path')
const { prepareMustache, setupMustache, mustacheFromEnv } = require('express-mustache-overlays')

const app = express()
prepareMustache(app, mustacheFromEnv(app))

// Any other express setup can change app.locals.mustache.libDirs here to add
// additional library-defined public files directories to be served.  Any user
// defined directories will be prepended before any corresponding URL path in
// the library directories list. The safest way to add overlays is with the
// overlay() function demonstrated here.
app.locals.mustache.overlay([path.join(__dirname, 'mustache')])

// Add any routes here:
app.get('', (req, res) => {
  res.render('hello', {})
})

// Set up the engine
const mustacheEngine = setupMustache(app)
app.engine('mustache', mustacheEngine)
app.set('views', app.locals.mustache.dirs)
app.set('view engine', 'mustache')

app.listen((process.env.PORT || 8000), () => console.log(`Example app listening on port ${(process.env.PORT || 8000)}`))


lib/lib.js
const mustache = require('mustache')
const path = require('path')
const debug = require('debug')('express-mustache-overlays')
const fs = require('fs')
const shelljs = require('shelljs')
const chokidar = require('chokidar')
const _ = require('lodash')
const { promisify } = require('util')

const readFileAsync = promisify(fs.readFile)
const accessAsync = promisify(fs.access)

const optionsFromEnv = (options) => {
  const { extended = true } = options || {}
  const mustacheDirs = process.env.MUSTACHE_DIRS ? process.env.MUSTACHE_DIRS.split(':') : []
  const publicFilesDirs = process.env.PUBLIC_FILES_DIRS ? process.env.PUBLIC_FILES_DIRS.split(':') : []
  const scriptName = process.env.SCRIPT_NAME || ''
  const publicUrlPath = process.env.PUBLIC_URL_PATH
  let result = { scriptName, publicUrlPath, mustacheDirs, publicFilesDirs }
  if (extended) {
    // These ones are used by in the default base template, that works in
    // conjunction with gateway-lite, or are commonly used, but they aren't strictly needed.
    const port = process.env.PORT || 80
    const sharedPublicUrlPath = process.env.SHARED_PUBLIC_URL_PATH
    const startUrl = process.env.START_URL
    const title = process.env.DEFAULT_TITLE
    const withPjaxPwa = (process.env.WITH_PJAX_PWA || 'false').toLowerCase() === 'true'
    const networkErrorUrl = process.env.NETWORK_ERROR_URL
    const manifestUrl = process.env.MANIFEST_URL
    const serviceWorkerUrl = process.env.SERVICE_WORKER_URL
    let icon192Url = process.env.ICON_192_URL
    const themeColor = process.env.THEME_COLOR
    result = Object.assign({}, result, { port, sharedPublicUrlPath, title, startUrl, withPjaxPwa, networkErrorUrl, manifestUrl, serviceWorkerUrl, icon192Url, themeColor })
  }
  return result
}

const loadPartials = async (mustacheDirs) => {
  const partials = {}
  if (!Array.isArray(mustacheDirs)) {
    throw Error('Expected an array of paths for mustachDirs')
  }
  mustacheDirs.reverse()
  for (let mustacheDir of mustacheDirs) {
    // Make it the same format that shelljs returns
    mustacheDir = path.normalize(mustacheDir)
    const partialsPath = path.join(mustacheDir, 'partials', '*.mustache')
    const files = shelljs.ls(partialsPath)
    if (shelljs.error()) {
      debug(`Error listing mustache partial files matching '${partialsPath}', perhaps none were found`)
    }
    if (!files.length) {
      debug('No partials found. Continuing anyway ...')
    }
    for (const filename of files) {
      const name = filename.slice((mustacheDir + '/partials/').length, filename.length - 9)
      partials[name] = await readFileAsync(filename, { encoding: 'utf8' })
    }
  }
  mustacheDirs.reverse()
  // Should match the file extension of the view.
  return partials
}

const mustacheOverlays = (locals) => {
  const { mustacheDirs } = locals || {}
  const overlaysPromise = new Promise(async (resolve, reject) => {
    debug('Setting up templates...')
    debug('Using these mustache views directories in order of preference:', mustacheDirs)
    let partials = await loadPartials(mustacheDirs)
    const partialsDirs = []
    const partialsDirsMustacheGlobs = []
    for (let mustacheDir of mustacheDirs) {
      partialsDirs.push(path.join(mustacheDir, 'partials'))
      partialsDirsMustacheGlobs.push(path.join(mustacheDir, 'partials', '**', '*.mustache'))
    }
    // Look for changes to partials directories, and reload if needed
    debug('Watching these globs:', partialsDirsMustacheGlobs.join(','))
    const onEvent = async (event, path) => {
      debug(`Reloading partials due to ${event} on ${path} ...`)
      partials = await loadPartials(mustacheDirs)
    }
    const throttledEvent = _.throttle(onEvent, 200, { 'trailing': true, 'leading': false })
    chokidar.watch(partialsDirsMustacheGlobs, { ignoreInitial: true, ignored: /(^|[/\\])\../ }).on('all', throttledEvent)
    const renderFile = async (filePath, options) => {
      const content = await readFileAsync(filePath, { encoding: 'utf8' })
      const rendered = mustache.render(content, Object.assign({}, options), partials)
      return rendered
    }
    const findView = async (template) => {
      let templatePath
      for (let i = 0; i < mustacheDirs.length; i++) {
        const filePath = path.join(mustacheDirs[i], template + '.mustache')
        try {
          await accessAsync(filePath, fs.constants.R_OK)
          templatePath = filePath
          break
        } catch (e) {
          // Can't access the file
          // debug(e)
        }
      }
      if (!templatePath) {
        throw new Error(`Template '${template}' not found, searched ${mustacheDirs}`)
      }
      return templatePath
    }
    const renderView = async (template, options) => {
      return renderFile((await findView(template)), options)
    }
    const overlays = {
      renderFile,
      renderView,
      findView
    }
    resolve(overlays)
  })
  return overlaysPromise
}

const mustacheEngine = (overlaysPromise) => {
  return async (filePath, options, callback) => {
    try {
      const overlays = await overlaysPromise
      const rendered = await overlays.renderFile(filePath, options)
      callback(null, rendered)
    } catch (e) {
      debug('Error setting up mustache overlays:', e)
      callback(e, null)
      throw new Error(e)
    }
  }
}

const installSignalHandlers = () => {
  // Better handling of SIGINT and SIGTERM for docker
  process.on('SIGINT', function () {
    console.log('Received SIGINT. Exiting ...')
    process.exit()
  })

  process.on('SIGTERM', function () {
    console.log('Received SIGTERM. Exiting ...')
    process.exit()
  })
}

const setupErrorHandlers = (app, options) => {
  const { debug = console.log, ...rest } = options || {}
  if (Object.keys(rest).length) {
    throw new Error(`Unexpected options passed to setupErrorHandlers(): ${Object.keys(rest).join(', ')}.`)
  }

  // Must be after other routes - Handle 404
  app.get('/*', (req, res) => {
    res.status(404)
    res.render('404')
  })

  // Error handler has to be last
  app.use('/*', function (err, req, res, next) {
    debug(err)
    res.status(500)
    try {
      res.render('500')
    } catch (e) {
      debug('Error during rendering 500 page:', e)
      res.send('Internal server error.')
    }
  })
}

module.exports = { optionsFromEnv, mustacheOverlays, mustacheEngine, installSignalHandlers, setupErrorHandlers }
